### PR TITLE
[Core,iOS,Android] SelectedItemColor in ListView

### PR DESCRIPTION
**Pull request by ravinderjangra**
### Description of Change ###

Added code for TextCell, ImageCell, SwitchCell, EntryCell, ViewCell to set BackgroundColor for SelectedItems in ListView.
Example:
```c#
 var t = new DataTemplate(()=> { return new ImageCell { SelectedBackgroundColor = Color.Red }; });
``` 
or in XAML
```xaml
  <TextCell SelectedBackgroundColor="Teal">
```

listview selected item color

### Bugs Fixed ###

No Bug Fix

### API Changes ###

Added:
 - Color SelectedBackgroundColor { get; set; } //Bindable Property in Cell

### Behavioral Changes ###

User will be able to set SelectedItem Background Color whole Defining the ItemTemplate.

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [ ] Changes adhere to coding standard
- [ ] Consolidate commits as makes sense

### CODE REVIEW

**Comment by ravinderjangra** [comment]
Is it going good?

**Comment by jassmith** [comment]
Only works on iOS, if the intention is to keep it that way it needs to be moved into a platformspecific API.

**Comment by ravinderjangra** [comment]
@jassmith  I will upload android soon. I committed to confirm if it the right way to apply this.

**Comment by jassmith** [comment]
We may want to rename the property but overall Im okay with this. There are some things you will need to change in your implementation.

File: Xamarin.Forms.Platform.iOS/Cells/EntryCellRenderer.cs [hunk 1]
```diff
@@ -23,6 +23,11 @@ public override UITableViewCell GetCell(Cell item, UITableViewCell reusableCell,
 				tvc.KeyboardDoneButtonPressed -= OnKeyBoardDoneButtonPressed;
 			}
 
+			tvc.SelectedBackgroundView = new UIView()
```
Comment (jassmith) [inline Xamarin.Forms.Platform.iOS/Cells/EntryCellRenderer.cs]: Do not set this unless item.SelectedBackgroundColor is set to something OTHER than Color.Default. If it gets set back to Color.Default it needs to be restored to its proper color. Also I don't see you doing change tracking anywhere which is a problem.

Comment (ravinderjangra) [reply Xamarin.Forms.Platform.iOS/Cells/EntryCellRenderer.cs]: @jassmith I was trying to work on tracking the change and updating the value but It is not hitting the HandlePropertyChanged. Can you suggest me why?

Comment (ravinderjangra) [reply Xamarin.Forms.Platform.iOS/Cells/EntryCellRenderer.cs]: and also already existing code for monitoring the change in imagesource in ImageCell is also not working.

Comment (jassmith) [reply Xamarin.Forms.Platform.iOS/Cells/EntryCellRenderer.cs]: we should fix that then...

Get the Default issue fixed first.

File: Xamarin.Forms.Core/Cells/Cell.cs [hunk 2]
```diff
@@ -13,6 +13,18 @@ public abstract class Cell : Element, ICellController, IFlowDirectionController
 		public const int DefaultCellHeight = 40;
 		public static readonly BindableProperty IsEnabledProperty = BindableProperty.Create("IsEnabled", typeof(bool), typeof(Cell), true, propertyChanged: OnIsEnabledPropertyChanged);
 
+		public static readonly BindableProperty SelectedItemBackgroundColorProperty =
+		BindableProperty.Create(nameof(SelectedItemBackgroundColor), typeof(Color), typeof(Cell), Color.Default);
+
+		/// <summary>
```
Comment (jassmith) [inline Xamarin.Forms.Core/Cells/Cell.cs]: Remove code docs

**Comment by jassmith** [comment]
build --uitest

**Comment by jassmith** [comment]
build --uitests

**Comment by jassmith** [comment]
always forget the s...

**Comment by elv1s42** [comment]
Hi All,

Any news about the release date?

Thank you

**Comment by PureWeen** [comment]
build --uitests

**Comment by kingces95** [comment]
Wrote a unit test and couldn't get it to work. Will circle back with the team tomorrow to see if maybe I'm holding it wrong.

**Comment by rmarinho** [comment]
@kingces95 might need to update the UITest package to the dev version (prerelease) to work locally

**Comment by ravinderjangra** [comment]
@kingces95 I'll check and post soon.

**Comment by rmarinho** [comment]
I had some code in my Windows machine as I was trying to do the UWP part, I looking now and seem I didn't stash that :( I clearly remember testing this on iOS while I was looking at the UWP bits. I think I followed the instructions example on the PR description. gonna try again today .

**Comment by ravinderjangra** [comment]
I think recent changes in listview broke this. I'll make required changes soon.

**Comment by rmarinho** [comment]
@kingces95 I did this here
https://github.com/xamarin/Xamarin.Forms/blob/9fe5372c1d80ca8fcff0c4fc397d25f91f6b61d3/Xamarin.Forms.Controls/CoreGallery.cs#L393

`var template = new DataTemplate(() => { return new TextCell { SelectedItemBackgroundColor = Color.Red }; });`

simulator screen shot - iphone x - 2018-09-12 at 14 58 08

screen shot 2018-09-12 at 15 03 28


On Android once you tap and old it doesn't get Red, but if you select it (just tap) you can see gets red, I wonder if we should also override and show the Red even when is not yet selected like tap and hold.

**Comment by jassmith** [comment]
build --uitests

**Comment by rmarinho** [comment]
build --uitests

**Comment by kingces95** [comment]
@ravinderjangra I'd like to include my test in your PR but I do not have permissions to push to your fork. I think you either have to add me to your fork or you have to add my changes to your fork yourself. My changes are in a branch `listview-bg-test3`.

Please try my test, and verify that it works or let me know what I'm doing wrong. As of yet, I'm still unable to see the colors change.

**Comment by samhouts** [comment]
@ravinderjangra Can you give us push permissions to this branch? https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/ Thanks!

**Comment by awatertrevi** [comment]
This does not work when setting the ListView's CachingStrategy to RecycleElement.

**Comment by samhouts** [comment]
@ravinderjangra Just checking in to see if you've had a chance to work on this. If you'd like for us to take it over, let us know. Thanks!

**Comment by ravinderjangra** [comment]
@samhouts thanks for checking in. I'm not getting enough time to work on this. So it will be great if you guys can take over this. Thanks.

**Comment by samhouts** [comment]
@ravinderjangra Will do. Did you give us push permissions to this branch? https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/ Thanks!

**Comment by kingces95** [comment]
Yea! Happy day. Pushed the UITest.

**Comment by samhouts** [comment]
build --uitests

**Comment by kingces95** [comment]
Please reopen with a UITest or verification that my test does not accurately stress the intended feature.

**Comment by lavilaso** [comment]
hi @samhouts @rmarinho , any news of this? this has the t / enhancement label from March 8, 2018 and the F100 label from March 20, 2018 and today May 12, 2019 this feature does not appear in the roadmap which is referenced until September 2019

**Comment by Syed-Esqimo** [comment]
Any updates on this? @samhouts and @kingces95

**Comment by samhouts** [comment]
@lavilaso @Syed-Esqimo This approach proved to be difficult for many edge cases. For CollectonView, we're recommending that you use the VisualStateManager to adjust the selected item color, and that will be our approach for ListView as well. This requires that https://github.com/xamarin/Xamarin.Forms/pull/5034 is finished, and it may also require that we add additional events to ListView that make it easier to leverage the VSM. So, it may look like this is not on the roadmap, but it is part of #5034, which _is_ on the roadmap. Thank you for your patience!

**Comment by titoleru** [comment]
This is so necessary... Custom renders do not work properly for Android :(

**Comment by titoleru** [comment]
I do not understand why is this issue closed at the moment. Sounds crazy to me...

**Comment by lavilaso** [comment]
> No entiendo por qué este problema está cerrado en este momento. A mí me parece una locura ...

Ultimately, great features are addressed in heaps, when you should first solve smaller things like these which torpedo productivity when working with the framework.

**Comment by matibak** [comment]
Hi @samhouts,
#8144 was closed, but this issue hasn't been resolved. Could you provide us the current status?

**Comment by titoleru** [comment]
Why was this issue closed??

**Comment by icecubetray** [comment]
bump

**Comment by samhouts** [comment]
As stated in https://github.com/xamarin/Xamarin.Forms/pull/2050#issuecomment-515121820, we recommend that you use VisualStateManager to achieve this. Please see https://github.com/hartez/SelectedItemStyling for examples. Thanks!

**Comment by matibak** [comment]
Will it work with ListView?

Den mån 8 juni 2020 21:46Samantha Houts <notifications@github.com> skrev:

**Pull request closed without merging**